allennlp/semparse/worlds/atis_world.py

```diff
@@ -0,0 +1,110 @@
+    ----------
+    utterances: ``List[str]``
+        A list of utterances in the interaction, the last element in this list is the
+        current utterance that we are interested in.
+    """
+    sql_table_context = SqlTableContext(TABLES)
+
+    def __init__(self, utterances: List[str], tokenizer=None) -> None:
+        self.utterances: List[str] = utterances
+        self.tokenizer = tokenizer if tokenizer else WordTokenizer()
+        self.tokenized_utterances = [self.tokenizer.tokenize(utterance) for utterance in self.utterances]
+        self.valid_actions: Dict[str, List[str]] = self.init_all_valid_actions()
+        self.grammar_str: str = self.get_grammar_str()
+        self.grammar_with_context = Grammar(self.grammar_str)
+
+    def get_valid_actions(self) -> Dict[str, List[str]]:
+        return self.valid_actions
+
+    def init_all_valid_actions(self) -> Dict[str, List[str]]:
+        """
+        We initialize the world's valid actions with that of the context. This means that the strings
+        and numbers that were valid earlier in the interaction are also valid. We then add new valid strings
+        and numbers from the current utterance.
+        """
+        valid_actions = deepcopy(self.sql_table_context.valid_actions)
+        for string in self.get_strings_from_utterance():
+            action = format_action('string', string)
+            if action not in valid_actions['string']:
+                valid_actions['string'].append(action)
+
+        numbers = ['0', '1']
+        for utterance in self.utterances:
+            numbers.extend(get_numbers_from_utterance(utterance))
+            for number in numbers:
+                action = format_action('number', number)
+                if action not in valid_actions['number']:
+                    valid_actions['number'].append(action)
+
+        return valid_actions
+
+    def get_grammar_str(self) -> str:
+        """
+        Generate a string that can be used to instantiate a ``Grammar`` object. The string is a sequence of
+        rules that define the grammar.
+        """
+        grammar_str_with_context = self.sql_table_context.grammar_str
+        numbers = [number.split(" -> ")[1].lstrip('["').rstrip('"]') for \
+                   number in sorted(self.valid_actions['number'], reverse=True)]
+        strings = [string .split(" -> ")[1].lstrip('["').rstrip('"]') for \
+                   string in sorted(self.valid_actions['string'], reverse=True)]
+
+        grammar_str_with_context += generate_one_of_string("number", numbers)
+        grammar_str_with_context += generate_one_of_string("string", strings)
+        return grammar_str_with_context
+
+
+    def get_strings_from_utterance(self) -> List[str]:
+        """
+        Based on the current utterance, return a list of valid strings that should be added.
+        """
+        strings: List[str] = []
+
+        for tokenized_utterance in self.tokenized_utterances:
+            if tokenized_utterance:
```

Comment:
Hmm, after thinking about this more, isn't the right way to handle an empty context to just have `self.utterances` and `self.tokenized_utterances` be empty lists?  In that case, you won't actually get into the body of this for loop, so you won't need this check.  Right?

Reply:
Right, I think that makes sense.